pydantic/generics.py

```diff
@@ -99,15 +107,48 @@ def __concrete_name__(cls: Type[Any], params: Tuple[Type[Any], ...]) -> str:
         """
         This method can be overridden to achieve a custom naming scheme for GenericModels
         """
-        param_names = [param.__name__ if hasattr(param, '__name__') else str(param) for param in params]
+        try:
+            GenericAlias = typing.GenericAlias  # type: ignore
+        except AttributeError:
+            GenericAlias = ()  # for all python < 3.9
+        param_names = [
+            param.__name__ if hasattr(param, '__name__') and not isinstance(param, GenericAlias) else str(param)
+            for param in params
+        ]
         params_component = ', '.join(param_names)
         return f'{cls.__name__}[{params_component}]'
 
 
-def resolve_type_hint(type_: Any, typevars_map: Dict[Any, Any]) -> Type[Any]:
-    if get_origin(type_) and getattr(type_, '__parameters__', None):
-        concrete_type_args = tuple([typevars_map[x] for x in type_.__parameters__])
-        return type_[concrete_type_args]
+def is_identity_typevars_map(typevars_map: Dict[TypeVarType, Type[Any]]) -> bool:
+    return bool(typevars_map) and all(k is v for k, v in typevars_map.items())
+
+
+def resolve_type_hint(type_: Any, typevars_map: Dict[Any, Any]) -> Any:
```

Comment:
docstring please.

Also are you sure this doesn't overlap with any of the methods in `pydantic.typing`? Could it be moved into `pydantic.typing`? Or doe it need to be here.

Reply:
I looked through `pydantic.typing` again and there's nothing in there that does this.  

The code is quite geared towards the `generics` module (i.e. I don't see anywhere else in pydantic where this would be used), but the code itself is relatively generic in the sense that it just recursively replaces types according to a a map/dictionary.

It does have to reference the `GenericModel` though so I think I would keep it in there until there is somewhere else that needs this sort of recursive type map.

I also though there may have been confusion regarding the `resolve_annotations` in the `pydantic.typing` module, so I renamed this to `replace_types` since that's what it does (not really specific to type hints)